src/dosext/dpmi/memory.c

```diff
@@ -237,6 +238,8 @@ static int SetAttribsForPage(unsigned int ptr, us attr, us old_attr)
     D_printf("Addr=%#x\n", ptr);
 
     if (change) {
+      if (config.cpu_vm_dpmi == CPUVM_EMU)
```

Comment:
Oh, do we need these checks everywhere?
How about moving them to the `e_invalidate_full()`
itself?

Reply:
ok. I was mostly thinking of the $_cpu_emu="vm86" case here with native DPMI. One can call e_invalidate_full unconditionally and it just won't do anything if cpuemu is not active for DPMI.
A little optimization is then as follows:
```
--- a/src/emu-i386/simx86/trees.c
+++ b/src/emu-i386/simx86/trees.c
@@ -1294,7 +1294,8 @@ void e_invalidate(unsigned data, int cnt)
  * Otherwise use e_invalidate() */
 void e_invalidate_full(unsigned data, int cnt)
 {
-       if (config.cpuemu <= 1)
+       if ((config.cpu_vm_dpmi != CPUVM_EMU) &&
+           (config.cpu_vm != CPUVM_EMU || data > LOWMEM_SIZE + HMASIZE))
                return;
        e_munprotect(data, cnt);
 #ifdef HOST_ARCH_X86
```
or maybe the inverse is more clear?
```
+        if (!((config.cpu_vm_dpmi == CPUVM_EMU) ||
+               (config.cpu_vm == CPUVM_EMU && data <= LOWMEM_SIZE + HMASIZE)))
```

Reply:
Why to remove `config.cpuemu` check?
It may be SIM, no?
Note that I completely removed trees from SIM
mode, so now calling them from SIM mode just
segfaults.
As for comparing address to lowmem_size -
I am neutral. It doesn't look like the obviously
right thing to do, but I can't suggest something
else either. :) Of course in theory it would be
nice if we use avl tree for mpmap. Then we
won't need that kind of an optimization, as the
code will be able to instantly see if the address
is out of range. But it seems like our mpmap
is just a single-linked list, so maybe indeed you
can save a few cpu cycles by doing that...
Maybe even a comment will be good that this
check is there because the mpmap look-ups
are sub-optimal.